embark-consult.el

```diff
@@ -181,6 +181,37 @@ This function is meant to be added to `embark-collect-mode-hook'."
       (when (fboundp 'wgrep-setup) (wgrep-setup)))
     (pop-to-buffer buf)))
 
+(defun embark-consult-export-lines-to-grep (lines)
+  "Create a grep mode buffer listing LINES as alternative to occur.
+This enables use of `wgrep' for editing of results including multi-line editing.
+Because `consult-line' produces grep-like results, the use of `wgrep' helps to
+increase uniformity of user experience when editing the results.
+The elements of LINES are assumed to be values of category `consult-line'."
+  (let ((buf (generate-new-buffer "*Embark Export Grep*"))
+        last-buf)
+    (with-current-buffer buf
+      (dolist (line lines)
+        (pcase-let*
+            ((`(,loc . ,num) (consult--get-location line))
+             (lineno (propertize (format "%d:" num)))
+             (contents (propertize (embark-consult--strip line)))
+             (this-buf (marker-buffer loc)))
+          (unless (eq this-buf last-buf)
+            (insert (propertize "Exported grep results:\n\n" 'wgrep-header t))
+            (setq last-buf this-buf))
+          (insert (concat (file-name-nondirectory (buffer-file-name this-buf)) ":" lineno contents "\n"))))
+      (goto-char (point-min))
+      (grep-mode)
+      (setq next-error-last-buffer buf)
+      ;; Set up keymap before possible wgrep-setup, so that wgrep
+      ;; restores our binding too when the user finishes editing.
+      (use-local-map (make-composed-keymap
+                      embark-consult-revert-map
+                      (current-local-map)))
+      (setq-local wgrep-header/footer-parser #'ignore)
+      (when (fboundp 'wgrep-setup) (wgrep-setup)))
```

Comment:
Is there some code here which could be shared with `embark-consult-export-grep`? Maybe one could introduce a helper function `embark-consult--export-grep` which sets up the grep buffer is passed a function as argument. The function argument should insert the actual content into the buffer.

~~~ elisp
(defun embark-consult--export-grep (insert)
  (let ((buf (generate-new-buffer "*Embark Export Grep*")))
    (with-current-buffer buf
      (insert (propertize "Exported grep results:\n\n" 'wgrep-header t))
      (funcall insert)
      (goto-char (point-min))
      (grep-mode)
      ;; Make this buffer current for next/previous-error
      (setq next-error-last-buffer buf)
      ;; Set up keymap before possible wgrep-setup, so that wgrep
      ;; restores our binding too when the user finishes editing.
      (use-local-map (make-composed-keymap
                      embark-consult-revert-map
                      (current-local-map)))
      (setq-local wgrep-header/footer-parser #'ignore)
      (when (fboundp 'wgrep-setup) (wgrep-setup)))
    (pop-to-buffer buf)))

(defun embark-consult-export-grep (lines)
  ...)
~~~

Reply:
I think that instead of the function you propose @minad, that this new export function should just build a grep-formatted list of lines and then call the existing `embark-consult-export-grep`.

Reply:
I thought a callback would be more efficient, since we can write directly to the buffer. Building an intermediate list requires allocations, which can be costly if we have many candidates. It would be even worse to build many intermediate strings.

Reply:
You're right, of course. I just thought it shouldn't be a problem since I'd expect few matching lines compared to the number of lines in the buffer, so the slow step would actually be the initial gathering of candidates in consult-line. I mostly wanted to reuse more code, in particular to reuse the line insertion loop, so maybe the best option is a callback but one that is called per line.